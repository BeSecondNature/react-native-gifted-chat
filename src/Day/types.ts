import {
  StyleProp,
  ViewStyle,
  TextStyle,
  TextProps,
} from 'react-native'
import { IMessage } from '../Models'

export interface DayProps<TMessage extends IMessage = IMessage> {
  currentMessage: TMessage
  nextMessage?: TMessage
  previousMessage?: TMessage
  containerStyle?: StyleProp<ViewStyle>
  wrapperStyle?: StyleProp<ViewStyle>
  textStyle?: StyleProp<TextStyle>
  textProps?: TextProps
  dateFormat?: string
  inverted?: boolean
}
