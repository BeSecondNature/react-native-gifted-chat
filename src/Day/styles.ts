import { StyleSheet } from 'react-native'
import Color from '../Color'

export default StyleSheet.create({
  container: {
    marginTop: 5,
    marginBottom: 10,
  },
  text: {
    backgroundColor: Color.backgroundTransparent,
    color: Color.defaultColor,
    fontSize: 12,
    fontWeight: '600',
  },
})
