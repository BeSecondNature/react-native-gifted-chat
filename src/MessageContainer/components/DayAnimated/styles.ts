import { StyleSheet } from 'react-native'
import Color from '../../../Color'

export default StyleSheet.create({
  dayAnimated: {
    position: 'absolute',
    width: '100%',
  },
  dayAnimatedDayContainerStyle: {
    marginTop: 0,
    marginBottom: 0,
  },
  dayAnimatedText: {
    fontSize: 12,
    color: Color.white,
  },
})
