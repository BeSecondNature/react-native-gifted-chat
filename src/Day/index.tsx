import * as React from 'react'
import {
  Text,
  View,
} from 'react-native'
import dayjs from 'dayjs'

import { isSameDay } from '../utils'
import { DATE_FORMAT } from '../Constant'
import { IMessage } from '../Models'
import { DayProps } from './types'

import { useChatContext } from '../GiftedChatContext'
import styles from './styles'

export * from './types'

export function Day<TMessage extends IMessage = IMessage> ({
  dateFormat = DATE_FORMAT,
  currentMessage,
  previousMessage,
  containerStyle,
  wrapperStyle,
  textStyle,
}: DayProps<TMessage>) {
  const { getLocale } = useChatContext()

  if (currentMessage == null || isSameDay(currentMessage, previousMessage))
    return null

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={wrapperStyle}>
        <Text style={[styles.text, textStyle]}>
          {dayjs(currentMessage.createdAt)
            .locale(getLocale())
            .format(dateFormat)}
        </Text>
      </View>
    </View>
  )
}
