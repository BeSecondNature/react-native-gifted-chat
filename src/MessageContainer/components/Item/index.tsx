import React, { forwardRef, useCallback, useMemo } from 'react'
import { LayoutChangeEvent, View } from 'react-native'
import { IMessage } from '../../../types'
import Message, { MessageProps } from '../../../Message'
import Animated, { interpolate, useAnimatedStyle, useDerivedValue, useSharedValue } from 'react-native-reanimated'
import { DaysPositions } from '../../types'
import { Day } from '../../../Day'
import { isSameDay } from '../../../utils'
import { ItemProps } from './types'

export * from './types'

export const useScrolledPosition = (listHeight: { value: number }, scrolledY: { value: number }, containerHeight: { value: number }, dayBottomMargin: number, dayTopOffset: number) => {
  const scrolledPosition = useDerivedValue(() =>
    listHeight.value + scrolledY.value - containerHeight.value - dayBottomMargin - dayTopOffset
  , [listHeight, scrolledY, containerHeight, dayBottomMargin, dayTopOffset])

  return scrolledPosition
}

export const useTopOffset = (
  listHeight: { value: number },
  scrolledY: { value: number },
  daysPositions: { value: DaysPositions },
  containerHeight: { value: number },
  dayBottomMargin: number,
  dayTopOffset: number
) => {
  const scrolledPosition = useScrolledPosition(listHeight, scrolledY, containerHeight, dayBottomMargin, dayTopOffset)

  const daysPositionsArray = useDerivedValue(() => Object.values(daysPositions.value).sort((a, b) => a.y - b.y))

  const currentDayPosition = useDerivedValue(() =>
    daysPositionsArray.value.find((day, index) => {
      const dayPosition = day.y + day.height
      return (scrolledPosition.value < dayPosition) || index === daysPositionsArray.value.length - 1
    })
  , [daysPositionsArray, scrolledPosition])

  const topOffset = useDerivedValue(() => {
    const scrolledBottomY = listHeight.value + scrolledY.value - (currentDayPosition.value?.height ?? 0) - 5 // 5 is marginTop in Day component.
    const dateBottomY = currentDayPosition.value?.y ?? 0

    const topOffset = scrolledBottomY - dateBottomY

    return topOffset
  }, [listHeight, scrolledY, currentDayPosition])

  return topOffset
}

const DayWrapper = forwardRef<View, MessageProps<IMessage>>((props, ref) => {
  const {
    renderDay: renderDayProp,
    currentMessage,
    previousMessage,
  } = props

  if (!currentMessage?.createdAt || isSameDay(currentMessage, previousMessage))
    return null

  const {
    /* eslint-disable @typescript-eslint/no-unused-vars */
    containerStyle,
    onMessageLayout,
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ...rest
  } = props

  return (
    <View ref={ref}>
      {
        renderDayProp
          ? renderDayProp({ ...rest, createdAt: currentMessage.createdAt })
          : <Day {...rest} createdAt={currentMessage.createdAt} />
      }
    </View>
  )
})

const Item = (props: ItemProps) => {
  const {
    onRefDayWrapper,
    renderMessage: renderMessageProp,
    scrolledY,
    daysPositions,
    listHeight,
    ...rest
  } = props

  const containerHeight = useSharedValue(0)
  const dayTopOffset = useMemo(() => 10, [])
  const dayBottomMargin = useMemo(() => 10, [])
  const topOffset = useTopOffset(listHeight, scrolledY, daysPositions, containerHeight, dayBottomMargin, dayTopOffset)

  const handleLayout = useCallback(({ nativeEvent }: LayoutChangeEvent) => {
    containerHeight.value = nativeEvent.layout.height
  }, [containerHeight])

  const style = useAnimatedStyle(() => ({
    opacity: interpolate(
      topOffset.value,
      [-dayTopOffset, -0.0001, 0, containerHeight.value + dayTopOffset],
      [0, 0, 1, 1],
      'clamp'
    ),
  }), [topOffset, containerHeight, dayTopOffset])

  const handleRef = useCallback((ref: any) => {
    onRefDayWrapper(
      ref,
      props.currentMessage._id,
      props.currentMessage.createdAt instanceof Date
        ? props.currentMessage.createdAt.getTime()
        : props.currentMessage.createdAt
    )
  }, [onRefDayWrapper, props.currentMessage])

  return (
    <View key={props.currentMessage._id.toString()}>
      <Animated.View style={style} onLayout={handleLayout}>
        <DayWrapper
          {...rest as MessageProps<IMessage>}
          ref={handleRef}
        />
      </Animated.View>
      {
        renderMessageProp
          ? renderMessageProp(rest as MessageProps<IMessage>)
          : <Message {...rest as MessageProps<IMessage>} />
      }
    </View>
  )
}

export default Item
