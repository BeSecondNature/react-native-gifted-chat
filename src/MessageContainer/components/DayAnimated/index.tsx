import React, { useCallback, useEffect, useMemo, useState } from 'react'
import Animated, { interpolate, useAnimatedStyle, useDerivedValue, useSharedValue, useAnimatedReaction, withTiming, runOnJS } from 'react-native-reanimated'
import { Day } from '../../../Day'
import { isSameDay } from '../../../utils'
import { useScrolledPosition, useTopOffset } from '../Item'
import { DayAnimatedProps } from './types'

import stylesCommon from '../../../styles'
import styles from './styles'

export * from './types'

const DayAnimated = ({ scrolledY, daysPositions, listHeight, renderDay, messages, isLoadingEarlier, ...rest }: DayAnimatedProps) => {
  const opacity = useSharedValue(0)
  const fadeOutOpacityTimeoutId = useSharedValue<ReturnType<typeof setTimeout> | null>(null)
  const containerHeight = useSharedValue(0)

  const isScrolledOnMount = useSharedValue(false)
  const isLoadingEarlierAnim = useSharedValue(isLoadingEarlier)

  const daysPositionsArray = useDerivedValue(() => Object.values(daysPositions.value).sort((a, b) => a.y - b.y))

  const [createdAt, setCreatedAt] = useState<number | undefined>()

  const dayTopOffset = useMemo(() => 10, [])
  const dayBottomMargin = useMemo(() => 10, [])
  const scrolledPosition = useScrolledPosition(listHeight, scrolledY, containerHeight, dayBottomMargin, dayTopOffset)
  const topOffset = useTopOffset(listHeight, scrolledY, daysPositions, containerHeight, dayBottomMargin, dayTopOffset)

  const messagesDates = useMemo(() => {
    const messagesDates: number[] = []

    for (let i = 1; i < messages.length; i++) {
      const previousMessage = messages[i - 1]
      const message = messages[i]

      if (!isSameDay(previousMessage, message) || !messagesDates.includes(new Date(message.createdAt).getTime()))
        messagesDates.push(new Date(message.createdAt).getTime())
    }

    return messagesDates
  }, [messages])

  const createdAtDate = useDerivedValue(() => {
    for (let i = 0; i < daysPositionsArray.value.length; i++) {
      const day = daysPositionsArray.value[i]
      const dayPosition = day.y + day.height - containerHeight.value - dayBottomMargin

      if (scrolledPosition.value < dayPosition)
        return day.createdAt
    }

    return messagesDates[messagesDates.length - 1]
  }, [daysPositionsArray, scrolledPosition, messagesDates, containerHeight, dayBottomMargin])

  const style = useAnimatedStyle(() => ({
    top: interpolate(
      topOffset.value,
      [-dayTopOffset, -0.0001, 0, isLoadingEarlierAnim.value ? 0 : containerHeight.value + dayTopOffset],
      [dayTopOffset, dayTopOffset, -containerHeight.value, isLoadingEarlierAnim.value ? -containerHeight.value : dayTopOffset],
      'clamp'
    ),
  }), [topOffset, containerHeight, dayTopOffset, isLoadingEarlierAnim])

  const contentStyle = useAnimatedStyle(() => ({
    opacity: opacity.value,
  }), [opacity])

  const fadeOut = useCallback(() => {
    'worklet'

    opacity.value = withTiming(0, { duration: 500 })
  }, [opacity])

  const scheduleFadeOut = useCallback(() => {
    if (fadeOutOpacityTimeoutId.value)
      clearTimeout(fadeOutOpacityTimeoutId.value)

    fadeOutOpacityTimeoutId.value = setTimeout(fadeOut, 500)
  }, [fadeOut, fadeOutOpacityTimeoutId])

  const handleLayout = useCallback(({ nativeEvent }) => {
    containerHeight.value = nativeEvent.layout.height
  }, [containerHeight])

  useAnimatedReaction(
    () => [scrolledY.value, daysPositionsArray],
    (value, prevValue) => {
      if (!isScrolledOnMount.value) {
        isScrolledOnMount.value = true
        return
      }

      if (value[0] === prevValue?.[0])
        return

      opacity.value = withTiming(1, { duration: 500 })

      runOnJS(scheduleFadeOut)()
    },
    [scrolledY, scheduleFadeOut, daysPositionsArray]
  )

  useAnimatedReaction(
    () => createdAtDate.value,
    (value, prevValue) => {
      if (value && value !== prevValue)
        runOnJS(setCreatedAt)(value)
    },
    [createdAtDate]
  )

  useEffect(() => {
    isLoadingEarlierAnim.value = isLoadingEarlier
  }, [isLoadingEarlierAnim, isLoadingEarlier])

  if (!createdAt)
    return null

  return (
    <Animated.View
      style={[stylesCommon.centerItems, styles.dayAnimated, style]}
      onLayout={handleLayout}
    >
      <Animated.View
        style={contentStyle}
        pointerEvents='none'
      >
        {
          renderDay
            ? renderDay({ ...rest, createdAt })
            : <Day
              {...rest}
              containerStyle={[styles.dayAnimatedDayContainerStyle, rest.containerStyle]}
              createdAt={createdAt}
            />
        }
      </Animated.View>
    </Animated.View>
  )
}

export default DayAnimated
