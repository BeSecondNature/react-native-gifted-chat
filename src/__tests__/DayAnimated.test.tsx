import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { DayAnimated } from '../MessageContainer/components/DayAnimated'
import { DayProps } from '../Day'

const mockDaysPositions = { value: {} }
const mockScrolledY = { value: 0 }
const mockListHeight = { value: 800 }

const mockMessage = {
  _id: 1,
  text: 'Hello',
  createdAt: new Date('2023-01-01'),
  user: { _id: 1, name: 'User 1' }
}

describe('DayAnimated', () => {
  it('should render DayAnimated with default Day component', () => {
    const component = renderer.create(
      <DayAnimated
        scrolledY={mockScrolledY}
        daysPositions={mockDaysPositions}
        listHeight={mockListHeight}
        messages={[mockMessage]}
        isLoadingEarlier={false}
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should use custom renderDay when provided', () => {
    const customRenderDay = jest.fn((props: DayProps) => <div data-testid="custom-day">Custom Day: {props.createdAt}</div>)
    
    const component = renderer.create(
      <DayAnimated
        scrolledY={mockScrolledY}
        daysPositions={mockDaysPositions}
        listHeight={mockListHeight}
        messages={[mockMessage]}
        isLoadingEarlier={false}
        renderDay={customRenderDay}
      />
    )
    
    // Force render to trigger the renderDay call if there's a createdAt
    component.getInstance()
    
    // The custom renderDay function should be available in the component
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})