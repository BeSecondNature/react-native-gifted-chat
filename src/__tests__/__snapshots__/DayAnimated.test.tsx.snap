// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayAnimated should render DayAnimated with default Day component 1`] = `null`;

exports[`DayAnimated should use custom renderDay when provided 1`] = `null`;
